src/Uno.Extensions.Reactive/Collections/Tracking/CollectionAnalyzer.SourceEnumerator.cs

```diff
@@ -0,0 +1,216 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Uno;
+
+namespace Uno.Extensions.Collections.Tracking;
+
+partial class CollectionAnalyzer
+{
+	private class SourceEnumerator<TCollection>
+		where TCollection : IList
+	{
+		private readonly TCollection _snapshot;
+		private readonly IndexOf<TCollection> _indexOf;
+		private readonly IgnoredIndexCollection _ignored = new();
+
+		public SourceEnumerator(
+			TCollection snapshot,
+			IndexOf<TCollection> indexOf)
+		{
+			_snapshot = snapshot;
+			_indexOf = indexOf;
+		}
+
+		public int CurrentIndex { get; private set; } = -1;
+
+		public object? Current { get; private set; }
+
+		/// <summary>
+		/// Count of items that have been ignored by the enumerator
+		/// </summary>
+		public int Ignored { get; private set; }
+
+		public bool MoveNext()
+		{
+			var index = CurrentIndex;
+			while (_ignored.TryDequeue(++index))
+			{
+				Ignored++;
+			}
+
+			if (index >= _snapshot.Count)
+			{
+				CurrentIndex = -1;
+				Current = default(object);
+
+				return false;
+			}
+			else
+			{
+				CurrentIndex = index;
+				Current = _snapshot[index];
+
+				return true;
+			}
+		}
+
+		public void Ignore(int index)
+		{
+			if (index <= CurrentIndex)
+			{
+				throw new InvalidOperationException("You cannot ignore an index which was already enumerated.");
```

Comment:
Can the message be more actionable? (Don't use directed language)

Reply:
This is an internal safety check, like a `Debug.Assert` except I prefer to explicitly fail even in release as I cannot recover from this case and propagating an invalid change-set can drive the app to crash.